File: arm.64/frame/avalonia/src/bind.fs
namespace Pat

open System
open System.IO
open System.Runtime.InteropServices

module Binder =
    let bind (binaryPath: string) : Cabi option =
        let filename = Path.GetFileName binaryPath
        let skip = [ "c"; "cpp"; "rust"; "go"; "swift"; "haskell"; "zig"; "v" ]
        if List.contains filename skip then None
        else
            let mutable handle = IntPtr.Zero
            if not (NativeLibrary.TryLoad(binaryPath, &handle)) then None
            else
                let mutable morePtr = IntPtr.Zero
                let mutable pumpPtr = IntPtr.Zero
                let mutable lessPtr = IntPtr.Zero
                if NativeLibrary.TryGetExport(handle, "More", &morePtr) &&
                   NativeLibrary.TryGetExport(handle, "Pump", &pumpPtr) &&
                   NativeLibrary.TryGetExport(handle, "Less", &lessPtr) then
                    Some {
                        Name = filename
                        Path = binaryPath
                        Lib = handle
                        More = Marshal.GetDelegateForFunctionPointer<MoreFn>(morePtr)
                        Pump = Marshal.GetDelegateForFunctionPointer<PumpFn>(pumpPtr)
                        Less = Marshal.GetDelegateForFunctionPointer<LessFn>(lessPtr)
                    }
                else
                    NativeLibrary.Free handle
                    None

File: arm.64/frame/blazor/src/bind.cs
using System;
using System.IO;
using System.Runtime.InteropServices;

namespace Pat {
    public static class Binder {
        public static Cabi? Bind(string binaryPath) {
            string filename = Path.GetFileName(binaryPath);
            if (filename == "c" || filename == "cpp" || filename == "rust" ||
                filename == "go" || filename == "swift" || filename == "haskell" ||
                filename == "zig" || filename == "v") {
                return null;
            }

            if (!NativeLibrary.TryLoad(binaryPath, out IntPtr handle)) {
                return null;
            }

            if (!NativeLibrary.TryGetExport(handle, "More", out IntPtr morePtr) ||
                !NativeLibrary.TryGetExport(handle, "Pump", out IntPtr pumpPtr) ||
                !NativeLibrary.TryGetExport(handle, "Less", out IntPtr lessPtr)) {
                NativeLibrary.Free(handle);
                return null;
            }

            return new Cabi {
                Name = filename,
                Path = binaryPath,
                Lib = handle,
                More = Marshal.GetDelegateForFunctionPointer<MoreFn>(morePtr),
                Pump = Marshal.GetDelegateForFunctionPointer<PumpFn>(pumpPtr),
                Less = Marshal.GetDelegateForFunctionPointer<LessFn>(lessPtr)
            };
        }
    }
}
